trace/src/main/org/jetbrains/lincheck/trace/AddressIndex.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2025 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.lincheck.trace
+
+import java.nio.file.Files
+import kotlin.math.min
+
+internal sealed class AddressIndex {
+    var size: Int = 0
+        protected set
+
+    abstract operator fun get(index: Int): Long
+
+    abstract fun add(address: Long): AddressIndex
+    open fun finishWrite() {}
+
+    internal companion object {
+        fun create(): AddressIndex = NullAddressIndex()
+    }
+}
+
+private class NullAddressIndex: AddressIndex() {
+    override fun get(index: Int): Long = -1L
+
+    override fun add(address: Long): AddressIndex {
+        if (address < 0) {
+            size++
+            return this
+        }
+        var replacement: AddressIndex = MemoryAddressIndex()
+        repeat(size) {
+            replacement = replacement.add(-1)
+        }
+        return replacement.add(address)
+    }
+}
+
+// 1 MiB of memory, max
+private const val MAX_MEM_INDEX_SIZE = 1024 * 1024 / Long.SIZE_BYTES
+
+private class MemoryAddressIndex: AddressIndex() {
+    private var storage: LongArray = LongArray(16)
+
+    override fun get(index: Int): Long {
+        if (index !in 0..<size) throw IndexOutOfBoundsException("Index: $index, Size: $size")
+        return storage.get(index)
+    }
+
+    override fun add(address: Long): AddressIndex {
+        if (address < 0) return this
+
+        if (size == MAX_MEM_INDEX_SIZE) {
+            val fileIndex = MemoryMapAddressIndex(this.storage, size)
```

Comment:
Same suggestion as for `RangeIndex`. 
Also, here it would help to get rid of separate class `NullAddressIndex` as you can store `indexImpl: AddressIndex?`